File: examples/many.cpp
/**
 * @file    many.cpp
 * @brief   Example program for solving multiple tridiagonal systems in 2D using PaScaL_TDMA.
 * @details This example demonstrates solving tridiagonal systems in both x and y directions using
 *          a 2D MPI process topology. The program distributes RHS data, performs two sequential TDMA
 *          solves (x- then y-direction), and collects the solution at the root rank for error analysis.
 *
 * Usage:
 *      mpirun -n <num_procs> ./many_example <nx> <ny>
 *      - nx : Number of unknowns in x-direction (10 ≤ nx ≤ 10,000)
 *      - ny : Number of unknowns in y-direction (10 ≤ ny ≤ 10,000)
 */

#include <vector>
#include <cmath>
#include <random>
#include <mpi.h>
#include <string>
#include <stdexcept>
#include <iostream>

#include "pascal_tdma.hpp"
#include "comm_layout_2d.hpp"

constexpr double a_diag = 10.0;
constexpr double a_upper = -1.0;
constexpr double a_lower = -1.0;

using namespace PaScaL_TDMA;

void generate_rhs(DimArray<double>& d, DimArray<double>& x, int nx, int ny);
void distribute_rhs(DimArray<double>& d_sub, const DimArray<double>& d, const DomainLayout2D& dom,
                    const CommLayout2D& topo);
void collect_solution(DimArray<double>& d, const DimArray<double>& d_sub, const DomainLayout2D& dom,
                      const CommLayout2D& topo);

int main(int argc, char** argv) {
    MPI_Init(&argc, &argv);

    if (argc != 3) throw std::runtime_error("Usage: many_example <nx> <ny>");

    const int nx = std::stoi(argv[1]);
    const int ny = std::stoi(argv[2]);
    if (nx < 10 || nx > 10000) throw std::runtime_error("Recommendation: 10 <= nx <= 10,000");
    if (ny < 10 || ny > 10000) throw std::runtime_error("Recommendation: 10 <= ny <= 10,000");

    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    bool is_root = (rank == 0);

    int dims[2] = {0, 0};
    int period[2] = {0, 0};
    MPI_Dims_create(size, 2, dims);

    CommLayout2D topo(dims, period);
    DomainLayout2D dom(nx, ny, topo);

    const int nx_sub = dom.getParDimX();
    const int ny_sub = dom.getParDimY();
    topo.buildCommBufferInfo(nx_sub, ny_sub);

    DimArray<double> d, x;
    if (is_root) generate_rhs(d, x, nx, ny);

    DimArray<double> d_sub(nx_sub, ny_sub), x_sub(nx_sub, ny_sub);
    distribute_rhs(d_sub, d, dom, topo);

    // Solve in x-direction
    DimArray<double> a_sub, b_sub, c_sub;
    a_sub.assign(nx_sub, ny_sub, a_lower);
    b_sub.assign(nx_sub, ny_sub, a_diag);
    c_sub.assign(nx_sub, ny_sub, a_upper);

    PTDMAPlanMany px_many;
    px_many.create(nx_sub, ny_sub, topo.getCommX(), TDMAType::Standard);
    PTDMASolverMany::solve(px_many, a_sub, b_sub, c_sub, d_sub);
    px_many.destroy();

    // Solve in y-direction
    DimArray<double> d_sub_tr(ny_sub, nx_sub);
    for (int i = 0; i < nx_sub; i++)
        for (int j = 0; j < ny_sub; j++) d_sub_tr(j, i) = d_sub(i, j);

    a_sub.assign(ny_sub, nx_sub, a_lower);
    b_sub.assign(ny_sub, nx_sub, a_diag);
    c_sub.assign(ny_sub, nx_sub, a_upper);

    PTDMAPlanMany py_many;
    py_many.create(ny_sub, nx_sub, topo.getCommY(), TDMAType::Cyclic);
    PTDMASolverMany::solve(py_many, a_sub, b_sub, c_sub, d_sub_tr);
    py_many.destroy();

    for (int i = 0; i < nx_sub; i++)
        for (int j = 0; j < ny_sub; j++) d_sub(i, j) = d_sub_tr(j, i);

    collect_solution(d, d_sub, dom, topo);

    if (is_root) {
        double norm2 = 0.0;
        for (int i = 0; i < nx * ny; i++) norm2 += std::pow(d.getVector()[i] - x.getVector()[i], 2);
        std::cout << "Avg. RMS error = " << std::sqrt(norm2 / nx / ny) << std::endl;
    }

    MPI_Finalize();
    return 0;
}

void generate_rhs(DimArray<double>& d, DimArray<double>& x, int nx, int ny) {
    DimArray<double> a, b, c, y(nx, ny);
    a.assign(nx, ny, a_lower);
    b.assign(nx, ny, a_diag);
    c.assign(nx, ny, a_upper);
    x.resize(nx, ny);
    d.resize(nx, ny);

    std::mt19937 gen(std::random_device{}());
    std::uniform_real_distribution<> dis(0.0, 1.0);
    for (int i = 0; i < nx; i++)
        for (int j = 0; j < ny; j++) x(i, j) = dis(gen);

    // y = A_j * x
    for (int i = 0; i < nx; i++) {
        y(i, 0) = a(i, 0) * x(i, ny - 1) + b(i, 0) * x(i, 0) + c(i, 0) * x(i, 1);
        for (int j = 1; j < ny - 1; j++)
            y(i, j) = a(i, j) * x(i, j - 1) + b(i, j) * x(i, j) + c(i, j) * x(i, j + 1);
        y(i, ny - 1) = a(i, ny - 1) * x(i, ny - 2) + b(i, ny - 1) * x(i, ny - 1) + c(i, ny - 1) * x(i, 0);
    }

    // d = A_i * y
    for (int j = 0; j < ny; j++) d(0, j) = b(0, j) * y(0, j) + c(0, j) * y(1, j);
    for (int i = 1; i < nx - 1; i++)
        for (int j = 0; j < ny; j++)
            d(i, j) = a(i, j) * y(i - 1, j) + b(i, j) * y(i, j) + c(i, j) * y(i + 1, j);
    for (int j = 0; j < ny; j++)
        d(nx - 1, j) = a(nx - 1, j) * y(nx - 2, j) + b(nx - 1, j) * y(nx - 1, j);
}

void distribute_rhs(DimArray<double>& d_sub, const DimArray<double>& d, const DomainLayout2D& dom,
                    const CommLayout2D& topo) {
    std::vector<double> d_blk;
    const auto& cnt_x = topo.getCountX();
    const auto& cnt_y = topo.getCountY();
    const auto& cnt_all = topo.getCountAll();
    const auto& disp_x = topo.getDisplX();
    const auto& disp_y = topo.getDisplY();
    const auto& disp_all = topo.getDisplAll();
    const int npx = topo.getSizeX();
    const int npy = topo.getSizeY();
    const int rank = topo.getRank();

    const int nx_sub = dom.getParDimX();
    const int ny_sub = dom.getParDimY();
    const int n_sub = dom.getParDimXY();
    const int nxy = dom.getDimXY();

    if (rank == 0) {
        int idx = 0;
        d_blk.resize(nxy);
        for (int iblk = 0; iblk < npx; iblk++)
            for (int jblk = 0; jblk < npy; jblk++)
                for (int i = 0; i < cnt_x[iblk]; i++)
                    for (int j = 0; j < cnt_y[jblk]; j++)
                        d_blk[idx++] = d(i + disp_x[iblk], j + disp_y[jblk]);
    }

    std::vector<double> recv_blk(n_sub);
    MPI_Scatterv(d_blk.data(), cnt_all.data(), disp_all.data(), MPI_DOUBLE, recv_blk.data(), n_sub,
                 MPI_DOUBLE, 0, MPI_COMM_WORLD);

    for (int i = 0; i < nx_sub; i++)
        for (int j = 0; j < ny_sub; j++) d_sub(i, j) = recv_blk[i * ny_sub + j];
}

void collect_solution(DimArray<double>& d, const DimArray<double>& d_sub, const DomainLayout2D& dom,
                      const CommLayout2D& topo) {
    const auto& cnt_x = topo.getCountX();
    const auto& cnt_y = topo.getCountY();
    const auto& cnt_all = topo.getCountAll();
    const auto& disp_x = topo.getDisplX();
    const auto& disp_y = topo.getDisplY();
    const auto& disp_all = topo.getDisplAll();
    const int npx = topo.getSizeX();
    const int npy = topo.getSizeY();
    const int rank = topo.getRank();

    const int nx_sub = dom.getParDimX();
    const int ny_sub = dom.getParDimY();
    const int n_sub = dom.getParDimXY();
    const int nxy = dom.getDimXY();

    std::vector<double> d_blk_recv;
    if (rank == 0) d_blk_recv.resize(nxy);

    std::vector<double> d_blk_send(n_sub);
    for (int i = 0; i < nx_sub; i++)
        for (int j = 0; j < ny_sub; j++) d_blk_send[j + i * ny_sub] = d_sub(i, j);

    MPI_Gatherv(d_blk_send.data(), n_sub, MPI_DOUBLE, d_blk_recv.data(), cnt_all.data(),
                disp_all.data(), MPI_DOUBLE, 0, MPI_COMM_WORLD);

    if (rank == 0) {
        int idx = 0;
        for (int iblk = 0; iblk < npx; iblk++)
            for (int jblk = 0; jblk < npy; jblk++)
                for (int i = 0; i < cnt_x[iblk]; i++)
                    for (int j = 0; j < cnt_y[jblk]; j++)
                        d(i + disp_x[iblk], j + disp_y[jblk]) = d_blk_recv[idx++];
    }
}

File: examples/fortran/many.f90
!===================================================================================================
!> @file        many.f90
!> @brief       Example program for solving multiple tridiagonal systems in 2D using PaScaL_TDMA.
!> @details     This example distributes a 2D domain across MPI processes, builds subcommunicators
!>              for x- and y-directions, and solves the tridiagonal systems in both directions using
!>              the PaScaL_TDMA library. It computes the average norm2 error compared to the 
!>              reference solution generated on the root process.
!===================================================================================================

program main

    use mpi
    use iso_c_binding
    use fortran_interface
    use mpi_topology_2d

    implicit none

    integer :: nx = 100, ny = 100
    integer :: nx_sub, ny_sub, n_sub
    integer :: nprocs, myrank, ierr
    integer :: npx
    integer :: para_range_n
    integer, allocatable :: cnt_x(:), disp_x(:), cnt_y(:), disp_y(:), cnt_all(:), disp_all(:)
    logical :: is_root = .false.

    double precision, allocatable :: d(:,:), x(:,:)
    double precision, allocatable :: a_sub(:,:), b_sub(:,:), c_sub(:,:), d_sub(:,:), d_sub_tr(:,:)

    type(c_ptr) :: px_many, py_many   !< Plans for multiple tridiagonal systems

    call MPI_Init(ierr)
    call MPI_Comm_size(MPI_COMM_WORLD, nprocs, ierr)
    call MPI_Comm_rank(MPI_COMM_WORLD, myrank, ierr)

    if (myrank == 0) is_root = .true.

    call MPI_Dims_create(nprocs, 2, np_dim, ierr)
    period = [.false., .false.]
    call mpi_topology_make()

    npx = np_dim(0)
    nx_sub = para_range_n(1, nx, comm_1d_x%nprocs, comm_1d_x%myrank)
    ny_sub = para_range_n(1, ny, comm_1d_y%nprocs, comm_1d_y%myrank)
    n_sub = nx_sub * ny_sub

    call build_comm_info_array()

    ! Allocate and build reference solution on root
    if (is_root) then
        allocate(d(nx, ny), x(nx, ny))
        call build_global_coeff_array()
    end if

    call distribute_rhs_array()

    ! Solve in y-direction
    allocate(a_sub(nx_sub, ny_sub), b_sub(nx_sub, ny_sub), c_sub(nx_sub, ny_sub))
    a_sub = 1.0; b_sub = 2.0; c_sub = 1.0

    call pascal_tdma_plan_many_create(py_many, ny_sub, nx_sub, comm_1d_y%myrank, comm_1d_y%nprocs, &
                                      comm_1d_y%mpi_comm, 0)
    call pascal_tdma_many_solve(py_many, a_sub, b_sub, c_sub, d_sub)
    call pascal_tdma_plan_many_destroy(py_many)

    ! Solve in x-direction
    a_sub = 1.0; b_sub = 2.0; c_sub = 1.0
    allocate(d_sub_tr(ny_sub, nx_sub))
    d_sub_tr = transpose(d_sub)

    call pascal_tdma_plan_many_create(px_many, nx_sub, ny_sub, comm_1d_x%myrank, comm_1d_x%nprocs, &
                                      comm_1d_x%mpi_comm, 0)
    call pascal_tdma_many_solve(px_many, a_sub, b_sub, c_sub, d_sub_tr)
    call pascal_tdma_plan_many_destroy(px_many)

    d_sub = transpose(d_sub_tr)
    deallocate(d_sub_tr)

    call collect_solution_array()

    if (is_root) print *, "Avg. norm2 error = ", norm2(d - x) / nx / ny

    call dealloc_all()
    call mpi_topology_clean()
    call MPI_Finalize(ierr)

contains
    !-----------------------------------------------------------------------------------------------
    !> @brief Build count and displacement arrays for scatter/gather operations.
    !-----------------------------------------------------------------------------------------------
    subroutine build_comm_info_array()

        integer :: i

        allocate(cnt_x(np_dim(0)), cnt_y(np_dim(1)), cnt_all(nprocs))
        allocate(disp_x(np_dim(0)), disp_y(np_dim(1)), disp_all(nprocs))

        call MPI_Allgather(nx_sub, 1, MPI_INTEGER, cnt_x, 1, MPI_INTEGER, comm_1d_x%mpi_comm, ierr)
        call MPI_Allgather(ny_sub, 1, MPI_INTEGER, cnt_y, 1, MPI_INTEGER, comm_1d_y%mpi_comm, ierr)
        call MPI_Allgather(n_sub,  1, MPI_INTEGER, cnt_all, 1, MPI_INTEGER, MPI_COMM_WORLD, ierr)

        disp_x(1) = 0
        do i = 2, size(cnt_x)
            disp_x(i) = disp_x(i - 1) + cnt_x(i - 1)
        end do

        disp_y(1) = 0
        do i = 2, size(cnt_y)
            disp_y(i) = disp_y(i - 1) + cnt_y(i - 1)
        end do

        disp_all(1) = 0
        do i = 2, nprocs
            disp_all(i) = disp_all(i - 1) + cnt_all(i - 1)
        end do

    end subroutine build_comm_info_array

    !-----------------------------------------------------------------------------------------------
    !> @brief Build reference coefficients and right-hand side arrays on the root process.
    !-----------------------------------------------------------------------------------------------
    subroutine build_global_coeff_array()

        integer :: i, j
        double precision, allocatable :: a(:,:), b(:,:), c(:,:), y(:,:)

        allocate(a(nx, ny), b(nx, ny), c(nx, ny), y(nx, ny))
        a = 1.0; b = 2.0; c = 1.0; y = 0.0

        call random_number(x)

        ! y = A_x * x
        do j = 1, ny
            y(1, j) = b(1, j) * x(1, j) + c(1, j) * x(2, j)
            do i = 2, nx - 1
                y(i, j) = a(i, j) * x(i - 1, j) + b(i, j) * x(i, j) + c(i, j) * x(i + 1, j)
            end do
            y(nx, j) = a(nx, j) * x(nx - 1, j) + b(nx, j) * x(nx, j)
        end do

        ! d = A_y * y
        do i = 1, nx
            d(i, 1) = b(i, 1) * y(i, 1) + c(i, 1) * y(i, 2)
        end do
        do j = 2, ny - 1
            do i = 1, nx
                d(i, j) = a(i, j) * y(i, j - 1) + b(i, j) * y(i, j) + c(i, j) * y(i, j + 1)
            end do
        end do
        do i = 1, nx
            d(i, ny) = a(i, ny) * y(i, ny - 1) + b(i, ny) * y(i, ny)
        end do

        deallocate(a, b, c, y)

    end subroutine build_global_coeff_array

    !-----------------------------------------------------------------------------------------------
    !> @brief Distribute the right-hand side array to all processes.
    !-----------------------------------------------------------------------------------------------
    subroutine distribute_rhs_array()

        integer :: i, j, iblk
        double precision, allocatable :: d_blk(:)

        if (is_root) then
            allocate(d_blk(nx * ny))
            do iblk = 1, npx
                do j = 1, ny
                    do i = 1, cnt_x(iblk)
                        d_blk(i + (j - 1) * cnt_x(iblk) + disp_x(iblk) * ny) = &
                            d(i + disp_x(iblk), j)
                    end do
                end do
            end do
        else
            allocate(d_blk(0))
        end if

        allocate(d_sub(nx_sub, ny_sub))
        call MPI_Scatterv(d_blk, cnt_all, disp_all, MPI_DOUBLE_PRECISION, d_sub, n_sub, &
                          MPI_DOUBLE_PRECISION, 0, MPI_COMM_WORLD, ierr)
        deallocate(d_blk)

    end subroutine distribute_rhs_array

    !-----------------------------------------------------------------------------------------------
    !> @brief Collect the solution from all processes to the root process.
    !-----------------------------------------------------------------------------------------------
    subroutine collect_solution_array()

        integer :: i, j, iblk
        double precision, allocatable :: d_blk(:)

        if (is_root) then
            allocate(d_blk(nx * ny))
        else
            allocate(d_blk(0))
        end if

        call MPI_Gatherv(d_sub, n_sub, MPI_DOUBLE_PRECISION, d_blk, cnt_all, disp_all, &
                         MPI_DOUBLE_PRECISION, 0, MPI_COMM_WORLD, ierr)

        if (is_root) then
            do iblk = 1, npx
                do j = 1, ny
                    do i = 1, cnt_x(iblk)
                        d(i + disp_x(iblk), j) = &
                            d_blk(i + (j - 1) * cnt_x(iblk) + disp_x(iblk) * ny)
                    end do
                end do
            end do
        end if

        deallocate(d_blk)

    end subroutine collect_solution_array

    !-----------------------------------------------------------------------------------------------
    !> @brief Deallocate all allocated arrays.
    !-----------------------------------------------------------------------------------------------
    subroutine dealloc_all()
        if (is_root) then
            deallocate(d, x)
        end if
        deallocate(a_sub, b_sub, c_sub, d_sub)
        deallocate(cnt_x, disp_x, cnt_y, disp_y, cnt_all, disp_all)
    end subroutine dealloc_all
end program main

!---------------------------------------------------------------------------------------------------
!> @brief Compute the number of elements assigned to each process in a 1D partitioning.
!---------------------------------------------------------------------------------------------------
integer function para_range_n(n1, n2, nprocs, myrank) result(n)

    implicit none

    integer, intent(in) :: n1, n2, nprocs, myrank
    integer :: remainder

    n = int((n2 - n1 + 1) / nprocs)
    remainder = mod(n2 - n1 + 1, nprocs)
    if (remainder > myrank) n = n + 1

end function para_range_n